ir/src/main/scala/org/scalajs/core/ir/Serializers.scala

```diff
@@ -808,4 +830,58 @@ object Serializers {
       strings(input.readInt())
     }
   }
+
+  private class RewriteArgumentsTransformer extends Transformers.Transformer {
+    import RewriteArgumentsTransformer._
+
+    private[this] var paramToIndex: Map[String, Int] = _
+
+    def transformMethodDef(tree: MethodDef): MethodDef = {
+      /* Ideally, we would re-hash the new MethodDef here, but we cannot do
+       * that because it prevents the JS version of the tools to link.
+       * Since the hashes of exported methods are not used by our pipeline
+       * anyway, we simply put None.
+       */
+      val MethodDef(static, name, args, resultType, body) = tree
+      setupParamToIndex(args)
+      MethodDef(static, name, List(argumentsParamDef(tree.pos)),
+          resultType, transform(body, isStat = resultType == NoType))(
+          tree.optimizerHints, None)(tree.pos)
+    }
+
+    def transformConstructorExportDef(
+        tree: ConstructorExportDef): ConstructorExportDef = {
+      val ConstructorExportDef(name, args, body) = tree
+      setupParamToIndex(args)
+      ConstructorExportDef(name, List(argumentsParamDef(tree.pos)),
+          transformStat(body))(tree.pos)
+    }
+
+    private def setupParamToIndex(params: List[ParamDef]): Unit =
+      paramToIndex = params.map(_.name.name).zipWithIndex.toMap
+
+    private def argumentsParamDef(implicit pos: Position): ParamDef =
+      ParamDef(Ident(ArgumentsName), AnyType, mutable = false, rest = true)
+
+    private def argumentsRef(implicit pos: Position): VarRef =
+      VarRef(Ident(ArgumentsName))(AnyType)
+
+    override def transform(tree: Tree, isStat: Boolean): Tree = tree match {
+      case VarRef(Ident(name, origName)) =>
+        implicit val pos = tree.pos
+        paramToIndex.get(name).fold {
+          if (name == "arguments") argumentsRef
+          else tree
+        } { paramIndex =>
+          JSBracketSelect(argumentsRef, IntLiteral(paramIndex))
+        }
+
+      case _ =>
+        super.transform(tree, isStat)
+    }
+  }
+
+  private object RewriteArgumentsTransformer {
+    private final val ArgumentsName = "$arguments"
```

Comment:
Nope. The compiler can never emit an identifier which is `$` followed by a valid identifier that does not begin by `$`. This is the same reason why we can use `$imul` for `ScalaJS.imul` in isolated mode, for example. Basically all of these are reserved for link-time stuff (deserialization, optimizations, desugaring).